src/queue.cpp

```diff
@@ -380,12 +355,15 @@ cl_int cvk_command_queue::finish() {
     std::lock_guard<std::mutex> lock(m_lock);
 
     auto status = flush_no_lock();
+    if (status != CL_SUCCESS) {
+        return status;
+    }
 
-    if ((status == CL_SUCCESS) && (m_finish_event != nullptr)) {
-        m_finish_event->wait();
+    if (m_executor != nullptr) {
+        m_executor->wait_idle();
```

Comment:
This works because we're holding the queue lock but it prevents parallel waiting on one thread while another thread keeps enqueueing commands in another. We should look into enabling this in the future. And to be clear: it's not a change in behaviour but the "wait idle" approach might make this harder to achieve in the future. We can deal with this in a later change. I've created https://github.com/kpet/clvk/issues/467.